opt: Introduce v1 of the Optgen language compiler

andy-kimball opened the pull request:
Optgen is a domain-specific language (DSL) that provides an
intuitive syntax for defining, matching, and replacing nodes in a
target expression tree. The Optgen language is described more
fully in the doc.go file in the optgen/lang directory. Cockroach
will use Optgen to generate parts of the cost-based optimizer,
including optimizer expression tree construction and pattern-
based transformations of the tree.

This PR contains a scanner that breaks Optgen input files into
lexical tokens, a parser that assembles an abstract syntax tree
(AST) from the tokens, and a compiler that performs semantic
checks and creates a rudimentary symbol table. This PR does *not*
contain the optimizer code generator - that will come in a future
PR. However, the langgen tool does generate the Optgen's own AST
using the Optgen syntax (compiler bootstrapping).

The code in this PR is derived from the v4 opttoy code, with
several big improvements:

  - Error detection and reporting is significantly improved
  - Code comments and documentation is significantly improved
  - The AST is generated using the Optgen syntax rather than
    being hand-coded
  - The components have much better tests

Release note: None

andy-kimball committed:
opt: Introduce v1 of the Optgen language compiler

Optgen is a domain-specific language (DSL) that provides an
intuitive syntax for defining, matching, and replacing nodes in a
target expression tree. The Optgen language is described more
fully in the doc.go file in the optgen/lang directory. Cockroach
will use Optgen to generate parts of the cost-based optimizer,
including optimizer expression tree construction and pattern-
based transformations of the tree.

This PR contains a scanner that breaks Optgen input files into
lexical tokens, a parser that assembles an abstract syntax tree
(AST) from the tokens, and a compiler that performs semantic
checks and creates a rudimentary symbol table. This PR does *not*
contain the optimizer code generator - that will come in a future
PR. However, the langgen tool does generate the Optgen's own AST
using the Optgen syntax (compiler bootstrapping).

The code in this PR is derived from the v4 opttoy code, with
several big improvements:

  - Error detection and reporting is significantly improved
  - Code comments and documentation is significantly improved
  - The AST is generated using the Optgen syntax rather than
    being hand-coded
  - The components have much better tests

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21732)
<!-- Reviewable:end -->

tschottdorf commented:
That's some pretty impressive work.

---

Reviewed 31 of 31 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/opt/optgen/cmd/langgen/main.go, line 96 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cFmKxqYwopK9L8QoV:-L3cFmKxqYwopK9L8QoW:b8mvyu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/cmd/langgen/main.go#L96)):*
> ```Go
> // usage is a replacement usage function for the flags package.
> func usage() {
> 	fmt.Fprintf(os.Stderr, "LangGen generates the AST for the Optgen language.\n\n")
> ```

nit:

```go
fmt.Fprintf(os.Stderr, `Line one
Line two

Line four`)
```

is a little more ergonomical (take it or leave it).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3cNkY09g0L_BCqNFG-:blepr18)*
<!-- Sent from Reviewable.io -->

petermattis commented:
No emacs syntax highlighting?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3cPBB7kZziXXXvlvkL:b44cpx)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I only skimmed the PR to keep informed. The data driven tests are very neat.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3cQiIyeRIiKMoDr6xX:bs6osoh)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Agreed, very impressive!

Will take some time to go through everything.

Looks like TeamCity doesn't like the compiler bootstrapping. @benesch / @jordanlewis  what do you reccomend? Would making sure that `make generate` first builds `langgen` be a good solution? Alternatively, we could provide a version of langgen as a vendored dependency (?)

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/lang/compiler_test.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3b4SnOAEcVy0PA4tla:-L3b4SnOAEcVy0PA4tlb:bouu28v) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/compiler_test.go#L27)):*
> ```Go
> )
> 
> var _ = fmt.Println
> ```

what is this for? I see it in multiple files.. If it's to force the import, that can be done with `_ "fmt"` in the imports. (but I don't see why forcing an import of `fmt` would be useful?)

---

*[pkg/sql/opt/optgen/lang/doc.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cs2U4zebZljqgJiTQ:-L3cs2U4zebZljqgJiTR:badlhkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> 
> /*
> ```

This documentation is fantastic! Thanks!

---

*[pkg/sql/opt/optgen/lang/doc.go, line 204 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cqdA2uUgPrf5qOeN3:-L3cqdA2uUgPrf5qOeN4:bml6mrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L204)):*
> ```Go
>   (Select
>     $input:*
>     (Filter [ ... $item:(Subquery) ... ])
> ```

What happens in this case if there are multiple children that match the subquery? Will just the first child match? If the result of that matches the pattern again, would the rule then be applied again until it doesn't match anymore?

Looking at the opttoy-v4 generated code, I think only the first normalization rule that matches is applied?

I guess this aspect depends on whatever `_gen.go` code we hook up and not on the language itself (but this is worth mentioning).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3csbxWgyUjTe-UTc9q:b-bd4dae)*
<!-- Sent from Reviewable.io -->

rytaft commented:
+1 looks great so far!  I've reviewed a few files and will come back to do the others.

---

Reviewed 4 of 31 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cpJenTNt0DgVy7Tq0:-L3cpJenTNt0DgVy7Tq1:b69x98i) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go#L63)):*
> ```Go
> 	// Generate the expression type.
> 	if isValueType(define) {
> 		fmt.Fprintf(g.w, "type %s %s\n", exprType, g.translateType(sliceElementType(define)))
> ```

This seems backwards - why are you using `sliceElementType` when `isValueType` is true, and `valueType` when `isSliceType` is true?

---

*[pkg/sql/opt/optgen/cmd/langgen/main.go, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3crhfe4AeQxri-JfWg:-L3crOOGYqpToeieldkG:bd6hwu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/cmd/langgen/main.go#L106)):*
> ```Go
> 	fmt.Fprintf(os.Stderr, "The commands are:\n\n")
> 	fmt.Fprintf(os.Stderr, "\texprs  generate expression definitions and functions\n")
> 	fmt.Fprintf(os.Stderr, "\tops    generates operator definitions and functions\n")
> ```

[nit] generates -> generate (to be consistent with the exprs line above)

---

*[pkg/sql/opt/optgen/cmd/langgen/main.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3csa311lK5iNrRGpjb:-L3csa311lK5iNrRGpjc:berypuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/cmd/langgen/main.go#L125)):*
> ```Go
> 
> 	buf.WriteString("// Code generated by langgen; DO NOT EDIT.\n\n")
> 	fmt.Fprintf(&buf, "  package lang\n\n")
> ```

[nit] why is this indented?

---

*[pkg/sql/opt/optgen/lang/compiler.go, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cyDOch65Puv1mDl28:-L3cyDOddVuKvjb_Jyno:b-a47xok) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/compiler.go#L202)):*
> ```Go
> 
> // acceptRuleMatchExpr performs rewrites on expressions within the rule's match
> // expression.
> ```

I don't see where the rewrites are happening.  Could you add some more details to this comment? Seems like this function is mostly updating the state of the `ruleCompiler`...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3d1Dyrk09KuEIZXpyl:bryxaei)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 19 of 31 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/lang/compiler_test.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dSWa47KevHXd3Iwsk:-L3dSWa5F75v9XTM_aA0:b-l7bvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/compiler_test.go#L68)):*
> ```Go
> 
> 	if len(c.Errors()) != 1 || c.Errors()[0].Error() != "file not found" {
> 		t.Errorf("expected error, found: %v", c.Errors()[0])
> ```

`c.Errors()` might be nil, right?  `c.Errors()[0]` -> `c.Errors()`

---

*[pkg/sql/opt/optgen/lang/doc.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cs2U4zebZljqgJiTQ:-L3daqrSchccmKbtOoRc:b2ngthy) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L15)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This documentation is fantastic! Thanks!
</blockquote></details>

+1!

---

*[pkg/sql/opt/optgen/lang/doc.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dVgemgpyOARJKMM8N:-L3dVgemgpyOARJKMM8O:b6rbmv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L83)):*
> ```Go
> 
>   [<tag-1-name>, <tag-2-name>, ...]
>     define <name> {
> ```

[nit] remove indentation for define

---

*[pkg/sql/opt/optgen/lang/doc.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dX-riUXON-dMBw3ey:-L3dX-riUXON-dMBw3ez:b-pbweas) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L186)):*
> ```Go
> the child node. For example:
> 
>   (Concat $left:* "") => $left
> ```

This example is a bit confusing.  Is this showing that the match pattern `(Concat $left:* "")` is equivalent to `$left`, or is this a rule to replace `(Concat $left:* "")` with `$left`?

---

*[pkg/sql/opt/optgen/lang/doc.go, line 367 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3diKUNTG6nEXddDHd9:-L3diKUNTG6nEXddDHdA:bjim8hu) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L367)):*
> ```Go
>   match-expr     = match | STRING | match-not-expr | match-any-expr |
>                    match-list-expr
>   match-and-expr = match-not-expr ('&' match-and-expr)?
> ```

Why is `match-and-expr` defined in terms of `match-not-expr`? Should it be in terms of `match-expr`?

---

*[pkg/sql/opt/optgen/lang/doc.go, line 371 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dj2Uuw_credUCQ4yt:-L3dj2Uuw_credUCQ4yu:b-dy8le7) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L371)):*
> ```Go
>   match-invoke   = '(' invoke-name ref* ')'
>   invoke-name    = IDENT
>   match-list     = '[' '...' match-fields-arg '...' ']'
> ```

Are you going to define `match-fields-arg`?

---

*[pkg/sql/opt/optgen/lang/lang.opt, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dhzM_14Gz5DmMvynE:-L3dhzM_14Gz5DmMvynF:b-eum1er) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/lang.opt#L95)):*
> ```
> }
> 
> # type OpNamesExpr []OpName
> ```

Should this be `OpNameExpr`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  most of my comments are pretty minor - the biggest possible issue is in testdata/compiler. I may be wrong, but it seems that the expected result is not actually the desired result for the one of the test cases....

---

Reviewed 8 of 31 files at r1.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/lang/parser.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dvV1Ke7YRNyd4CUU3:-L3dvV1Ke7YRNyd4CUU4:b-19cszo) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/parser.go#L374)):*
> ```Go
> // match-and-expr = match-not-expr ('&' match-and-expr)?
> func (p *Parser) parseMatchAndExpr() Expr {
> 	left := p.parseMatchNotExpr()
> ```

Why is this `parseMatchNotExpr` instead of `parseMatchExpr`? Perhaps `match-not-expr` is a bit of a misnomer - maybe `match-bool-expr` might be more appropriate?

---

*[pkg/sql/opt/optgen/lang/scanner.go, line 309 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3e2J4NlY9fkNhs4IeT:-L3e2J4Omsl1g7Ffxebq:b35lv42) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/scanner.go#L309)):*
> ```Go
> 
> 	s.lineInfo.pos = s.lineInfo.prev
> 	s.lineInfo.prev = -1
> ```

what happens if you call `unread` twice in a row? Does `UnreadRune` handle it?

---

*[pkg/sql/opt/optgen/lang/testdata/compiler, line 175 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3e6ca1H_shbq2R60id:-L3e6ca1H_shbq2R60ie:b-wuuutu) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/testdata/compiler#L175)):*
> ```
> 						Label="input"
> 						Target=(Match
> 							Names=(OpNames SubOp1Op SubOp2Op)
> ```

Shouldn't this be expanded?

---

*[pkg/sql/opt/optgen/lang/testdata/compiler, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3e6r8VRRt2rKCorlnT:-L3e6r8VRRt2rKCorlnU:bo4s5xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/testdata/compiler#L183)):*
> ```
> 			Replace=(Construct
> 				OpName=(Construct
> 					OpName="OpName"
> ```

Shouldn't this be replaced with either `SubOp1Op` or `SubOp2Op`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3eAd2XDX6_ZJoupUkA:b-c2gwdb)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
The `make generate` problem is fixed by making this change:
```diff
+++ b/Makefile
@@ -356,7 +356,8 @@ $(BOOTSTRAP_TARGET): $(GITHOOKS) Gopkg.lock $(LOCAL_BIN)/returncheck | $(SUBMODU
 		./vendor/golang.org/x/perf/cmd/benchstat \
 		./vendor/golang.org/x/tools/cmd/goimports \
 		./vendor/golang.org/x/tools/cmd/goyacc \
-		./vendor/golang.org/x/tools/cmd/stringer
+		./vendor/golang.org/x/tools/cmd/stringer \
+		./pkg/sql/opt/optgen/cmd/langgen
```

You might need to `rm bin/.bootstrap` to force the target to rebuild. Note that running `make generate` actually generates some diffs, the generated files need to be updated in the commit.

---

Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3eWgbN6-IADxiNoVRn:b-pju7ru)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Thanks,  @radu. If I do the Makefile change, will others who sync and build Cockroach also need to remove the bootstrap file? Is there some convention for notifying others of changes like this that might require clean rebuilds?

---

Review status: 18 of 32 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cpJenTNt0DgVy7Tq0:-L3dpoioSNyQ6rvGz0OD:b-2mh1t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go#L63)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This seems backwards - why are you using `sliceElementType` when `isValueType` is true, and `valueType` when `isSliceType` is true?
</blockquote></details>

Yeah, it's backwards. It just so happens that they did the same thing (get the type of first field), so I didn't catch the bug. I fixed it and added an assert to make sure it doesn't happen again.

---

*[pkg/sql/opt/optgen/cmd/langgen/main.go, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3crhfe4AeQxri-JfWg:-L3dqjci-X4J_V-kocJN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/cmd/langgen/main.go#L106)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] generates -> generate (to be consistent with the exprs line above)
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/cmd/langgen/main.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3csa311lK5iNrRGpjb:-L3dqkbo-mOQ-vL04wNa:b-ep3a89) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/cmd/langgen/main.go#L125)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] why is this indented?
</blockquote></details>

Fixed. The go fmt tool was removing the indentation, so I didn't notice it.

---

*[pkg/sql/opt/optgen/lang/compiler.go, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cyDOch65Puv1mDl28:-L3drBnX0Sdm6QOe7lDo:b-z5v2cd) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/compiler.go#L202)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I don't see where the rewrites are happening.  Could you add some more details to this comment? Seems like this function is mostly updating the state of the `ruleCompiler`...
</blockquote></details>

It's just doing semantic checks at this point (was doing rewrites in the past). I updated the comment.

---

*[pkg/sql/opt/optgen/lang/compiler_test.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3b4SnOAEcVy0PA4tla:-L3drL-P0U6OC1aOgPzZ:bx3v21f) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/compiler_test.go#L27)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

what is this for? I see it in multiple files.. If it's to force the import, that can be done with `_ "fmt"` in the imports. (but I don't see why forcing an import of `fmt` would be useful?)
</blockquote></details>

I use fmt.Println to debug issues, so I often add that line, so that I don't have to add it every time I need to add a Println. I'll use _ "fmt" instead in the imports when I need it.

---

*[pkg/sql/opt/optgen/lang/compiler_test.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dSWa47KevHXd3Iwsk:-L3drxJOgeYCrNXyJAWG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/compiler_test.go#L68)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

`c.Errors()` might be nil, right?  `c.Errors()[0]` -> `c.Errors()`
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/lang/doc.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dVgemgpyOARJKMM8N:-L3dsEBjjeGIx0DBjYLp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L83)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] remove indentation for define
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/lang/doc.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dX-riUXON-dMBw3ey:-L3dsK072xfc8GWIYFhv:bcfzbz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L186)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This example is a bit confusing.  Is this showing that the match pattern `(Concat $left:* "")` is equivalent to `$left`, or is this a rule to replace `(Concat $left:* "")` with `$left`?
</blockquote></details>

It's intended to be a rule that expresses the identity that concatenating an empty string to a string expression can be simplified to the string expression itself. I changed the text to look like this, which is hopefully clearer?
```
  [EliminateConcat]
  (Concat $left:* "") => $left

If Concat's right operand is a literal string expression that's equal to the
empty string, then the pattern matches.
```

---

*[pkg/sql/opt/optgen/lang/doc.go, line 204 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3cqdA2uUgPrf5qOeN3:-L3dtaS7AV41KFCrhaz5:biwfaub) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L204)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What happens in this case if there are multiple children that match the subquery? Will just the first child match? If the result of that matches the pattern again, would the rule then be applied again until it doesn't match anymore?

Looking at the opttoy-v4 generated code, I think only the first normalization rule that matches is applied?

I guess this aspect depends on whatever `_gen.go` code we hook up and not on the language itself (but this is worth mentioning).
</blockquote></details>

The list matcher only considers the first child that matches. That would trigger the replace pattern. However, all the patterns are rematched against whatever the replace pattern produces, so if there are further matching children, they'd normally get picked up then. This "transitive matching" behavior is the reason why the replace pattern generally removes matching items from the list (to hoist them or push them down), or otherwise modifies them so that they don't match again. The memo does remember fingerprints that would break any "matching cycles" if they occur, but they really should not be occurring with our normalization patterns. The point of a normalization pattern is to drive towards a canonical form, but that doesn't make sense in the presence of a matching cycle.

I added additional text to the docs to clarify some of these points.

---

*[pkg/sql/opt/optgen/lang/doc.go, line 367 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3diKUNTG6nEXddDHd9:-L3eEc1EmFYmrjAWcGsC:b-pieu4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L367)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why is `match-and-expr` defined in terms of `match-not-expr`? Should it be in terms of `match-expr`?
</blockquote></details>

I made some changes in this area, since the naming wasn't quite right. The important thing is that only the first expression can match a child - the others are always custom matchers. See if my new naming is better.

---

*[pkg/sql/opt/optgen/lang/doc.go, line 371 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dj2Uuw_credUCQ4yt:-L3e4Xdf1CatlFmnkaHg:b1b7bz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L371)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Are you going to define `match-fields-arg`?
</blockquote></details>

Fixed this.

---

*[pkg/sql/opt/optgen/lang/lang.opt, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dhzM_14Gz5DmMvynE:-L3ennJr6ggmkkMunygo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/lang.opt#L95)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Should this be `OpNameExpr`?
</blockquote></details>

Done.

---

*[pkg/sql/opt/optgen/lang/parser.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dvV1Ke7YRNyd4CUU3:-L3enpL2yzMDHJuMlXMy:b-3tew03) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/parser.go#L374)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why is this `parseMatchNotExpr` instead of `parseMatchExpr`? Perhaps `match-not-expr` is a bit of a misnomer - maybe `match-bool-expr` might be more appropriate?
</blockquote></details>

See if my new naming is better.

---

*[pkg/sql/opt/optgen/lang/scanner.go, line 309 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3e2J4NlY9fkNhs4IeT:-L3enxTODdBzvnryR3sM:b-aj6w2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/scanner.go#L309)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

what happens if you call `unread` twice in a row? Does `UnreadRune` handle it?
</blockquote></details>

I added an assert to detect if unread is called twice. It never should be, because the scanner only looks ahead one character at most.

---

*[pkg/sql/opt/optgen/lang/testdata/compiler, line 175 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3e6ca1H_shbq2R60id:-L3eoXE5beROrhqrNKVp:b8838tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/testdata/compiler#L175)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Shouldn't this be expanded?
</blockquote></details>

For now, only matches at the top-level are expanded. Expanding matches down in subtrees like this can cause a combinatorial explosion, so before we did that we'd have to think through the implications. Right now the code generator will handle this case by generating code using ExprView rather than the  concrete expr types. It's a tad slower, but should still be pretty good.

---

*[pkg/sql/opt/optgen/lang/testdata/compiler, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3e6r8VRRt2rKCorlnT:-L3epSHanXrXoOdmGeJn:b-j1qjz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/testdata/compiler#L183)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Shouldn't this be replaced with either `SubOp1Op` or `SubOp2Op`?
</blockquote></details>

See my answer above. This case generates dynamic construction code rather than statically expanding the patterns.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3eqpiPKxWVQA77XpQ2:b-aiptgc)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
I addressed comments in my latest revision. I also tightened up how I handle source location information on the Optgen exprs. I changed the expression String method to print out source locations, and when I did that I discovered additional problems with source locations that I needed to fix.

---

Review status: 18 of 32 files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3erUQ1wJUQ_QyUfJLo:b35rk7w)*
<!-- Sent from Reviewable.io -->

benesch commented:
@andy-kimball @RaduBerinde I'm working on improvements to the Makefile to better handle binary dependencies at build-time. The patch @RaduBerinde suggested will work short-term, but it will require folks to manually remove .bin/bootstrap whenever langgen meaningfully changes. (Fine with me to merge with that patch, though; I'll send a followup PR if this merges first.)

Also, @andy-kimball, can we move the langgen command to `pkg/cmd/langgen`? (Writing the Makefile rules is easier when all of our commands live in the same place.)

---

Review status: 18 of 32 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3f-OS7A-kfS_XpHOoe:b87sku6)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
I hate to break up optgen into various far-flung directories, since it's a logically cohesive set of packages. What if I instead moved it into its own repo, maybe https://github.com/cockroachdb/optgen? I'd love to see the Cockroach codebase become less coupled and monolithic, and breaking out standalone functionality would be a step towards that goal.

---

Review status: 18 of 32 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3f2CT5rDe8c5iiyNyr:bq3b8sw)*
<!-- Sent from Reviewable.io -->

benesch commented:
> I'd love to see the Cockroach codebase become less coupled and monolithic, and breaking out standalone functionality would be a step towards that goal.

Heh, sounds like you and I are on the opposite ends of the monorepo debate. :)

I don't mean to wade in too deeply here, but two things to consider: other repos don't benefit from our static analysis (gofmt, crlfmt, go vet, errcheck, returncheck, etc.) and updating our vendor submodule is painful, especially when the code is in question is new and thus likely to change frequently.

---

Review status: 18 of 32 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3f4wbEBOqad1COF9XE:b-8sz6wn)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Yes, it's kind of a pain to update dependencies, my position would be to avoid that.

We could have a `pkg/cmd/langgen` that is just a wrapper that calls into the `optgen/cmd/langgen` package for the main functionality. (and that package could actually be moved up to `optgen/langgen`)

benesch commented:
👍🏽

I don’t feel strongly about moving it to pkg/cmd/langgen, though, as the
benefits to the build system as of right now are quite modest.

On Wed, Jan 24, 2018 at 11:28 PM RaduBerinde <notifications@github.com>
wrote:

> Yes, it's kind of a pain to update dependencies, my position would be to
> avoid that.
>
> We could have a pkg/cmd/langgen that is just a wrapper that calls into
> the optgen/cmd/langgen package for the main functionality. (and that
> package could actually be moved up to optgen/langgen)
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/21732#issuecomment-360357593>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IIH1AbTF6KDulsmRATgFocKinDX_ks5tOALUgaJpZM4Rqy0y>
> .
>

andy-kimball commented:
I'm actually fairly indifferent as to whether we have one repo or many - I'm mostly concerned with creating decoupled packages that don't "import the world", and that aren't broken up across the codebase in a way that makes it hard to see the boundaries. If the company has already made a decision to go with a monorepo, then I'm OK with that...though I do believe it's one reason that Google's technology works amazingly within Google, but terribly outside Google - if you put all code within convenient reach, then developers will use it, even when they shouldn't! That creates dependency spaghetti over time.

As for pkg/cmd, it appears that what's there now are mostly general-purpose tools that might be used in various places across the codebase. But langgen (and soon optgen) are specific-purpose tools that are only going to be used within sql/opt. I think we should establish a convention where general-purpose tools go in pkg/cmd, but specific-purpose tools go in an appropriate cmd directory further down the tree.

---

Review status: 18 of 32 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3hR2nrRXbvoahX9Qz6:bvqi6ru)*
<!-- Sent from Reviewable.io -->

benesch commented:
Ok, color me convinced on the pkg/cmd point!

FWIW, I don't think we've made an explicit decision to go with a monorepo,
and there are a few places where we’ve split components out. Our
arbitrary-precision decimals are separate (apd), as are a few glue packages
(go-libedit, strftime), plus our ORM adapters and ORM tests. In my opinion
this has been a mixed bag. A separate APD library is great for the Go
ecosystem, but divorcing our ORM tests from Cockroach causes an annoying
amount of version skew-related test failures. Others may feel differently.

I think it makes a lot of sense to make generally-useful packages we've
built accessible to the community, but wanted to point out that our tooling
isn't very conducive to it right now.

N

On Thu, Jan 25, 2018 at 8:21 AM Andy Kimball <notifications@github.com>
wrote:

> I'm actually fairly indifferent as to whether we have one repo or many -
> I'm mostly concerned with creating decoupled packages that don't "import
> the world", and that aren't broken up across the codebase in a way that
> makes it hard to see the boundaries. If the company has already made a
> decision to go with a monorepo, then I'm OK with that...though I do believe
> it's one reason that Google's technology works amazingly within Google, but
> terribly outside Google - if you put all code within convenient reach, then
> developers will use it, even when they shouldn't! That creates dependency
> spaghetti over time.
>
> As for pkg/cmd, it appears that what's there now are mostly
> general-purpose tools that might be used in various places across the
> codebase. But langgen (and soon optgen) are specific-purpose tools that are
> only going to be used within sql/opt. I think we should establish a
> convention where general-purpose tools go in pkg/cmd, but specific-purpose
> tools go in an appropriate cmd directory further down the tree.
> ------------------------------
>
> Review status: 18 of 32 files reviewed at latest revision, 18 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-:-L3hR2nrRXbvoahX9Qz6:bvqi6ru>*
>
> —
>
>
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/21732#issuecomment-360464471>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IMGKIQ7e6dsRFWg4rAwiIBjiyG73ks5tOH_BgaJpZM4Rqy0y>
> .
>

rytaft commented:
Reviewed 13 of 14 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go, line 362 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3i-4m5HOtjZ8Rc8s6z:-L3i-4m5HOtjZ8Rc8s7-:b4svzuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/07b8d54a659fe712457e03432239a3dc1fe672df/pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go#L362)):*
> ```Go
> // stores the original source information (file, line, pos).
> func hasSourceField(d *lang.DefineExpr) bool {
> 	return !isValueType(d) && !isSliceType(d) && len(d.Fields) > 0
> ```

In line 73 above it looks like you've added a source field for all non-value and non-slice types -- even when `len(d.Fields) == 0`.

---

*[pkg/sql/opt/optgen/lang/doc.go, line 186 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3dX-riUXON-dMBw3ey:-L3i1h0WOmlUKuvfEQhu:blgnl2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L186)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It's intended to be a rule that expresses the identity that concatenating an empty string to a string expression can be simplified to the string expression itself. I changed the text to look like this, which is hopefully clearer?
```
  [EliminateConcat]
  (Concat $left:* "") => $left

If Concat's right operand is a literal string expression that's equal to the
empty string, then the pattern matches.
```
</blockquote></details>

Looks good - thanks!

---

*[pkg/sql/opt/optgen/lang/doc.go, line 367 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3diKUNTG6nEXddDHd9:-L3i2sK8aIMKxgcRB0st:b-vs1pr4) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/doc.go#L367)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I made some changes in this area, since the naming wasn't quite right. The important thing is that only the first expression can match a child - the others are always custom matchers. See if my new naming is better.
</blockquote></details>

Looks good.

---

*[pkg/sql/opt/optgen/lang/testdata/compiler, line 175 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3e6ca1H_shbq2R60id:-L3i5oQBSlnJ8Aaob8XX:b-wsvasg) ([raw file](https://github.com/cockroachdb/cockroach/blob/677b948afcbe3d07d92ac6390b68e52f452fde2a/pkg/sql/opt/optgen/lang/testdata/compiler#L175)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

For now, only matches at the top-level are expanded. Expanding matches down in subtrees like this can cause a combinatorial explosion, so before we did that we'd have to think through the implications. Right now the code generator will handle this case by generating code using ExprView rather than the  concrete expr types. It's a tad slower, but should still be pretty good.
</blockquote></details>

Thanks for the explanation!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go, line 362 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732#-L3i-4m5HOtjZ8Rc8s6z:-L3nVn_nee2ED2681mQr:bg1ktfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/07b8d54a659fe712457e03432239a3dc1fe672df/pkg/sql/opt/optgen/cmd/langgen/exprs_gen.go#L362)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

In line 73 above it looks like you've added a source field for all non-value and non-slice types -- even when `len(d.Fields) == 0`.
</blockquote></details>

You're right. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21732)*
<!-- Sent from Reviewable.io -->
